File: Data Structure/python/three_sum_closest.py
from typing import List


class Solution:
    """
    Args:
        nums: Given an array of integers with negative value and duplicates.

    Returns:
        return the sum of the three integers is cloest to target. Assume that each input would have exactly one solution.

    Idea of Solution:
        Similar to 3 Sum problem, use 3 pointers to point current element, next element and the last element.
        If the sum is less than target, it means we have to add a larger element so next element move to the next.
        If the sum is greater, it means we have to add a smaller element so last element move to the second last element.
        Keep doing this until the end. Each time compare the difference between sum and target, 
        if it is less than minimum difference so far, then replace result with it, otherwise keep iterating.

    Time Complexity: O(n^2)
    Space Complexity: O(1)
    """

    def threeSumClosest(self, nums: List[int], target: int) -> int:
        nums.sort()
        res = float('inf')
        for i in range(len(nums) - 1):
            if i > 0 and nums[i] == nums[i-1]:
                continue
            # Two Pointer
            left, right = i + 1, len(nums) - 1
            while left < right:
                cloest = nums[i] + nums[left] + nums[right]
                if cloest == target:
                    return cloest
                if abs(target - cloest) < abs(target - res):
                    res = cloest
                if cloest > target:
                    right -= 1
                if cloest < target:
                    left += 1
        return res

File: Data Structure/java/ThreeSumClosest.java
import java.util.*;

public class ThreeSumClosest {
    public int threeSumClosest(int[] nums, int target) {
        Arrays.sort(nums);
        int res = Integer.MAX_VALUE - Math.abs(target); // We need to do this trick since Integer.MIN_VALUE always
                                                        // return negative in Math.abs
        for (int i = 0; i < nums.length - 2; i++) {
            if (i > 0 && nums[i] == nums[i - 1])
                continue;
            // Two Pointer
            int left = i + 1, right = nums.length - 1;
            while (left < right) {
                int sum = nums[i] + nums[left] + nums[right];
                if (sum == target)
                    return sum;
                if (Math.abs(target - sum) < Math.abs(target - res)) {
                    res = sum;
                }
                if (sum > target)
                    right--;
                if (sum < target)
                    left++;
            }
        }
        return res;
    }
}